### PR TITLE
Updated websocket consumers for ASGI v2.3. 

### PR DESCRIPTION
**Pull request by kristjanvalur**
This PR adds the "headers" field for `websocket.accept` messages and the "reason" field for `websocket.close`, which
were added in ASGI spec versions 2.1 and 2.3 respectively.

### CODE REVIEW

**Review by carltongibson — COMMENTED** [review]
Hi @kristjanvalur — thanks for this. Looks right. 

Can you make the tests pass? (Missing a `django_db` marker from the looks of it.) 


Refs https://asgi.readthedocs.io/en/latest/specs/www.html#websocket

**Comment by kristjanvalur** [comment]
I've updated, could you re-run the workflow?

**Comment by kristjanvalur** [comment]
Oops, sorry about that, everything should now pass, including the linting.

File: channels/generic/websocket.py [hunk 1]
```diff
@@ -44,11 +44,16 @@ def websocket_connect(self, message):
...
         self.accept()
 
-    def accept(self, subprotocol=None):
+    def accept(self, subprotocol=None, headers=None):
         """
         Accepts an incoming socket
         """
-        super().send({"type": "websocket.accept", "subprotocol": subprotocol})
+        message = {"type": "websocket.accept", "subprotocol": subprotocol}
+        if self.scope.get("spec_version", "2.0") >= "2.1":
```
Comment (Kludex) [inline channels/generic/websocket.py:52]: Is there a need for this kind of conditional?

Comment (kristjanvalur) [reply channels/generic/websocket.py:52]: I don't know, not sure if WSGI allows unspecified extra data in its messages.  A server with an older version will _probably_ just ignore it, but I wasn't sure.  What is the usual practice, do you think?

Comment (carltongibson) [reply channels/generic/websocket.py:52]: Good hustle on the asgiref ticket @kristjanvalur. Looks like we can drop this check? 

If you do that I'll pull this in. 

Thanks! 🎁

**Comment by kristjanvalur** [comment]
I did the fixes with github.dev, could not test them, but I´m crossing my fingers.

File: channels/generic/websocket.py [hunk 2]
```diff
@@ -79,14 +83,16 @@ def send(self, text_data=None, bytes_data=None, close=False):
         if close:
             self.close(close)
 
-    def close(self, code=None):
+    def close(self, code=None, reason=None):
         """
         Closes the WebSocket from the server end
         """
+        message = {"type": "websocket.close"}
         if code is not None and code is not True:
```
Comment (Kludex) [inline channels/generic/websocket.py:91]: For my curiosity, why is there a check for the "code not being True"?

File: channels/generic/websocket.py [hunk 3]
```diff
@@ -79,14 +83,16 @@ def send(self, text_data=None, bytes_data=None, close=False):
...
-            super().send({"type": "websocket.close", "code": code})
-        else:
-            super().send({"type": "websocket.close"})
+            message["code"] = code
+        if reason:
+            message["reason"] = reason
```
Comment (Kludex) [inline channels/generic/websocket.py:94]: I'm not sure if those conditionals are needed, but should be fine.

Comment (kristjanvalur) [reply channels/generic/websocket.py:94]: The idea is to not pollute the response with an empty "reason" which is optional in any case.

File: channels/generic/websocket.py (hunk 2 above)
Comment (kristjanvalur) [reply channels/generic/websocket.py:91]: Oh, hard questions!   I think this is a compatibility check.  It's been a while, but IIRC, there was a case where a "True" value was being passed to close...  I need to look.

Comment (kristjanvalur) [reply channels/generic/websocket.py:91]: Oh, this is not my code.  Phew.  So, your guess is as good as mine.  Maybe in the distant past, someone would call `close(True)` for whatever reason and this is trying to allow for that?  Nothing in the library or unit tests which is doing that, though.

**Pull request merged**